wasm/src/language_service.rs

```diff
@@ -1,40 +1,194 @@
+            log::trace!("found manifest at {manifest_dir:?}");
+
+            let manifest_dir = PathBuf::from(manifest_dir);
+
+            let exclude_files =
+                match js_sys::Reflect::get(&js_val, &JsValue::from_str("excludeFiles")) {
+                    Ok(v) => match v.dyn_into::<js_sys::Array>() {
+                        Ok(arr) => arr
+                            .into_iter()
+                            .filter_map(|x| x.as_string())
+                            .collect::<Vec<_>>(),
+                        Err(e) => unreachable!("controlled callback should have returned an array -- our typescript bindings should guarantee this. {e:?}"),
+                    },
+                    Err(_) => unreachable!("our typescript bindings should guarantee that an object with an excludeFiles property is returned here"),
+                };
+            let exclude_regexes =
+                match js_sys::Reflect::get(&js_val, &JsValue::from_str("excludeRegexes")) {
+                    Ok(v) => match v.dyn_into::<js_sys::Array>() {
+                        Ok(arr) => arr
+                            .into_iter()
+                            .filter_map(|x| x.as_string())
+                            .collect::<Vec<_>>(),
+                        Err(e) => unreachable!("controlled callback should have returned an array -- our typescript bindings should guarantee this. {e:?}"),
+                    },
+                    Err(_) => unreachable!("our typescript bindings should guarantee that an object with an excludeRegexes property is returned here"),
+                };
+
+            Some(ManifestDescriptor {
+                manifest: Manifest {
+                    exclude_regexes,
+                    exclude_files,
+                    ..Default::default()
+                },
+                manifest_dir,
+            })
+        };
+        let get_manifest = call_async_js_fn!(get_manifest, transformer);
+
+        let diagnostics_callback = to_js_function(diagnostics_callback.obj, "diagnostics_callback");
+
         let diagnostics_callback = diagnostics_callback
             .dyn_ref::<js_sys::Function>()
             .expect("expected a valid JS function")
             .clone();
-        let inner = qsls::LanguageService::new(move |update| {
-            let diags = update
-                .errors
-                .iter()
-                .map(|err| VSDiagnostic::from_compile_error(&update.uri, err))
-                .collect::<Vec<_>>();
-            let _ = diagnostics_callback
-                .call3(
-                    &JsValue::NULL,
-                    &update.uri.into(),
-                    &update.version.into(),
-                    &serde_wasm_bindgen::to_value(&diags)
-                        .expect("conversion to VSDiagnostic should succeed"),
-                )
-                .expect("callback should succeed");
-        });
+
+        let inner = qsls::LanguageService::new(
+            move |update| {
```

Comment:
I believe it would also help readability to move the diagnostics callback closure out of the call to `new` so that it's super clear to see how `inner` gets constructed and what the final `diagnostics_callback` does e.g.:
```rust
let diagnostics_callback = move |update| {
    let diags = update
        .errors
        .iter()
        .map(|err| VSDiagnostic::from_compile_error(&update.uri, err))
        .collect::<Vec<_>>();
    let _ = diagnostics_callback
        .call3(
            &JsValue::NULL,
            &update.uri.into(),
            &update.version.into(),
            &serde_wasm_bindgen::to_value(&diags)
                .expect("conversion to VSDiagnostic should succeed"),
        )
        .expect("callback should succeed");
};

let inner = qsls::LanguageService::new(
    diagnostics_callback,
    read_file,
    list_directory,
    get_manifest,
);
```